src/par_iter/internal.rs

```diff
@@ -69,6 +76,75 @@ pub trait UnindexedConsumer<ITEM>: Consumer<ITEM> {
     fn to_reducer(&self) -> Self::Reducer;
 }
 
+/// A splitter controls the policy for splitting into smaller work items.
+trait Splitter: Send + Sized {
+    fn needs_split(&self) -> bool;
+    fn split_off(self) -> (Self, Self);
+}
+
+/// Classic cost-splitting uses weights to split until below a threshold.
+#[derive(Clone, Copy)]
+struct CostSplitter(f64);
+
+impl Splitter for CostSplitter {
+    #[inline]
+    fn needs_split(&self) -> bool {
+        self.0 > THRESHOLD
+    }
+
+    #[inline]
+    fn split_off(mut self) -> (Self, Self) {
+        self.0 /= 2.0;
+        (self, self)
+    }
+}
+
+/// Thief-splitting is an adaptive policy that starts by splitting into enough
+/// jobs for every worker thread, and then resets itself whenever a job is
+/// actually stolen into a different thread.
+#[derive(Clone, Copy)]
+struct ThiefSplitter {
+    origin: usize,
+    splits: usize,
+}
+
+impl ThiefSplitter {
+    #[inline]
+    fn id() -> usize {
+        // The actual `ID` value is irrelevant.  We're just using its TLS
+        // address as a unique thread key, faster than a real thread-id call.
+        thread_local!{ static ID: bool = false; }
+        ID.with(|id| id as *const bool as usize )
```

Comment:
Ah, this is cute. I was grabbing the worker's thread index, but that was a bit annoying since sometimes we were not on a worker, so I had to force us to get into a worker before-hand. I was wondering how you avoided that.

(An interesting thing to think about also is whether we can incorporate some simple version of this into  `rayon::join` -- or probably a wrapper around it -- so that algorithms like quicksort could benefit.)